src/mscorlib/src/System/IO/Path.cs

```diff
@@ -151,28 +151,47 @@ public static class Path
             return null;
         }
 
-       
         // Returns the directory path of a file path. This method effectively
         // removes the last element of the given file path, i.e. it returns a
         // string consisting of all characters up to but not including the last
         // backslash ("\") in the file path. The returned value is null if the file
         // path is null or if the file path denotes a root (such as "\", "C:", or
         // "\\server\share").
-        //
-        public static String GetDirectoryName(String path) {
-            if (path != null) {
+        public static String GetDirectoryName(String path)
+        {
+            if (path != null)
+            {
                 CheckInvalidPathChars(path);
 
-                string normalizedPath = NormalizePath(path, false);
+                // Expanding short paths is dangerous in this case as the results will change with the current directory.
+                //
+                // Suppose you have a path called "PICTUR~1\Foo". Now suppose you have two folders on disk "C:\Mine\Pictures Of Me"
+                // and "C:\Yours\Pictures of You". If the current directory is neither you'll get back "PICTUR~1". If it is "C:\Mine"
+                // get back "Pictures Of Me". "C:\Yours" would give back "Pictures of You".
+                //
+                // Because of this and as it isn't documented that short paths are expanded we will not expand short names unless
+                // we're in legacy mode.
+                string normalizedPath = NormalizePath(path, fullCheck: false, expandShortPaths:
+#if FEATURE_PATHCOMPAT
+                    AppContextSwitches.UseLegacyPathHandling
+#else
+                    false
+#endif
+                );
 
                 // If there are no permissions for PathDiscovery to this path, we should NOT expand the short paths
                 // as this would leak information about paths to which the user would not have access to.
-                if (path.Length > 0)
+                if (path.Length > 0
+#if FEATURE_CAS_POLICY
+                    // Only do the extra logic if we're not in full trust
+                    && !CodeAccessSecurityEngine.QuickCheckForAllDemands()
+#endif
+                    )
                 {
                     try
                     {
                         // If we were passed in a path with \\?\ we need to remove it as FileIOPermission does not like it.
-                        string tempPath = Path.RemoveLongPathPrefix(path);
+                        string tempPath = RemoveLongPathPrefix(path);
```

Comment:
Hmm- seems to be. :/

Most of the code isn't hit as `System.Runtime.Extensions` is overriding most of this. One notable exception is the bring up of the AppDomain.

Generally speaking we're lacking on testing here. I'm trying to find a good way to test directly against CoreLib's mscorlib. We sanity check this stuff by running legacy tests internally against a non Core build (e.g. desktop) of this code- which clearly doesn't get the Core specific forks of the code.